include/lbann/weights/weights_proxy.hpp

```diff
@@ -0,0 +1,170 @@
+ *  Access is read-only; updates to the proxied weights must be
+ *  handled directly through the origin object.
+ *
+ *  I'm thinking about ways to enforce this without users having to
+ *  think (how horrible would that be!), but right now the contract
+ *  for this class is that the weights are read from the master copy
+ *  at the beginning of forward prop and then remain constant
+ *  throughout the iteration. At the end of the iteration, the proxy
+ *  values are considered invalid once gradient updates have begun.
+ *
+ *  @tparam TensorDataType The type to which the weights are proxied.
+ */
+template <typename TensorDataType>
+class WeightsProxy
+{
+  using DataTypeWeights = data_type_weights<TensorDataType>;
+  using ValuesType = El::AbstractDistMatrix<TensorDataType>;
+  using ValuesPtrType = std::unique_ptr<ValuesType>;
+public:
+
+  /** @brief Construct an empty proxy. */
+  WeightsProxy() = default;
+
+  /** @brief Construct a proxy given the master object.
+   *
+   *  @param w Master weights object, which must have a valid storage
+   *           matrix initialized internally.
+   */
+  WeightsProxy(weights const& w)
+    : master_weights_{&w},
+      values_{setup_values_(w)}
+  {}
+
+  /** @brief Copy a WeightsProxy object.
+   *
+   *  Creates a new WeightsProxy to the same object.
+   */
+  WeightsProxy(WeightsProxy const& other)
+    : WeightsProxy(other.master_weights_)
+  {}
+
+  /** @brief Move a WeightsProxy object. */
+  WeightsProxy(WeightsProxy&& other)
+    : master_weights_{other.master_weights_},
+      values_{std::move(other.values_)}
+  {}
+
+  WeightsProxy& operator=(WeightsProxy const& other)
+  {
+    this->setup(other.master_weights_);
+    return *this;
+  }
+
+  WeightsProxy& operator=(WeightsProxy&& other)
+  {
+    WeightsProxy<TensorDataType> tmp(std::move(other));
+    this->swap(tmp);
+    return *this;
+  }
+
+  /** @brief Check if the proxy is referencing a weights object. */
+  bool empty() const noexcept { return values_ == nullptr; }
+
+  /** @brief Provide setup function for delayed construction.
+   *
+   *  This overwrites any existing data.
+   *
+   *  @param w The weights object to be proxied.
+   */
+  void setup(weights const& w)
+  {
+    auto vals = setup_values_(w);
+    master_weights_ = &w;
+    std::swap(vals, values_);
+  }
+
+  /** @brief Swap contents with another WeightsProxy object. */
+  void swap(WeightsProxy<TensorDataType>& other)
+  {
+    std::swap(master_weights_, other.master_weights_);
+    std::swap(values_, other.values_);
+  }
+
+  /** @brief Synchronize the held values with the master set. */
+  void synchronize_with_master()
+  {
+    if (!values_->Viewing())
+      El::Copy(master_weights_->get_values(), *values_);
+  }
+
+  /** @brief Access the values. */
+  ValuesType const& values() const noexcept { return *values_; }
+
+  /** @brief Access the master weights object directly. */
+  weights const& master_weights() const noexcept { return *master_weights_; }
```

Comment:
```suggestion
  /** @brief Access the values. */
  ValuesType const& values() const { return *values_; }

  /** @brief Access the master weights object directly. */
  weights const& master_weights() const { return *master_weights_; }
```
Won't we hit an exception if the user requests the values before setup?